### ISSUE TITLE
"".to_string() is slower than String::new()

### ISSUE THREAD

**Issue by cgaebel** [rust-lang/rust#18404]
In a lot of rust code, I see `"".to_string()` being used as the default method of creating a new string. This would normally be fine, but it's actually noticeably slower than `String::new()` because it goes through the formatting infrastructure.

It'd be nice if this pattern could be special-cased in `String::to_string`'s implementation, or the optimizer made to do this transformation better.

**Comment by SimonSapin** [comment]
What is the preferred way to do this these days? https://github.com/servo/servo/pull/4682 is changing Servo to use `to_owned`, but it implies importing the `ToOwned` trait all over the place. Such imports for such a common operation should really be in the prelude.

**Comment by lfairy** [comment]
Note that we can't just special case `ToString` for `String`, since that would lead to an overlapping instance.

I have another idea though. How about we implement `to_string()` in `std::fmt` instead? By putting it in `std::fmt` we can deal with the internal `Formatter` directly, cutting out all that indirection.

That won't solve the problem of allocating too much, but it might lessen the bloat a bit.

**Comment by lfairy** [comment]
After some thought, I came up with another, simpler, idea.

First, add a method to `Display`:

```rust
pub trait Display {
    // ...
    #[inline]
    fn as_str_slice(&self) -> Option<&str> { None }
}
```

This method will return `None` for all types except `str`, where it returns `Some(self)`:

```rust
impl Display for str {
    // ...
    #[inline]
    fn as_str_slice(&self) -> Option<&str> { Some(self) }
}
```

Now, here's where the magic lies. In `ToString`, before invoking the format machinery we check `as_str_slice` first. If this returns `Some`, we simply copy the slice, skipping the machinery altogether:

```rust
impl<T: Display + ?Sized> ToString for T {
    fn to_string(&self) -> String {
        match self.as_str_slice() {
            Some(s) => s.to_owned(),
            None => // as before
        }
    }
}
```

Since `.as_str_slice()` is inlined, the branch should be optimized out (I'm not an expert on this, so correct me if I'm wrong). So it'll work the same as before for non-string types, but for `str` it'll use the fast `.to_owned()`.

Thoughts? I guess it's a bit icky having an extra method there, but if we hide it from the docs it should be unobtrusive enough.

**Comment by lfairy** [comment]
Okay, so I prototyped my idea. Initial benchmarks look very promising, with `"".to_string()` optimizing down to single call to `.to_owned()`, and `123.to_string()` almost unaffected by the change.

There's a problem though with smart pointer types, specifically `Arc<T>`. `Arc` implements `Display`, which gives it a `ToString` instance via the blanket impl. Problem is, if `T` is not a string, then `.to_string()` will now borrow the `Arc` *twice*: once when calling `.as_str_slice()`, and once when calling `format!()`.

To get around this issue, we can punt this branch onto the type level, using either associated types or negative bounds. Unfortunately the associated types solution doesn't work yet (#20400), and negative bounds are still in the pipeline (rust-lang/rfcs#586).

So I guess we'll have to wait.

**Comment by lfairy** [comment]
Okay, so I prototyped my idea. Initial benchmarks look very promising, with `"".to_string()` optimizing down to single call to `.to_owned()`, and `123.to_string()` almost unaffected by the change.

There's a problem though with smart pointer types, specifically `Arc<T>`. `Arc` implements `Display`, which gives it a `ToString` instance via the blanket impl. Problem is, if `T` is not a string, then `.to_string()` will now borrow the `Arc` *twice*: once when calling `.as_str_slice()`, and once when calling `format!()`. The compiler may be able to optimize that out, but it feels fragile.

This decision should really be made on the type level, using either associated types or negative bounds. Unfortunately the associated types solution doesn't work yet (#20400), and negative bounds are still in the pipeline (rust-lang/rfcs#586).

So I guess we'll have to wait.

**Comment by huonw** [comment]
> Problem is, if T is not a string, then .to_string() will now borrow the Arc twice: once when calling .as_str_slice(), and once when calling format!().

Borrowing a pointer out of an `Arc` is very cheap (only a pointer offset compared to a plain `Box` or `&`); the expensive part is construction/copying (`.clone()`) and destruction when the reference counts are modified atomically.

**Comment by lfairy** [comment]
> Borrowing a pointer out of an `Arc` is very cheap (only a pointer offset compared to a plain `Box` or `&`); the expensive part is construction/copying (`.clone()`) and destruction when the reference counts are modified atomically.

Ah, I stand corrected! That makes my solution practical, though it still feels like a hack.

I guess if we treat this as a temporary solution, to be removed when negative bounds land, it should be okay.

**Comment by ranma42** [comment]
cc me

**Comment by ahmedcharles** [comment]
Is this still relevant/reproducible?

**Comment by frewsxcv** [comment]
Specialization has landed. This can be worked on now?

**Comment by japaric** [comment]
@frewsxcv there's already a PR: #32586

**Comment by apasel422** [comment]
Was this effectively fixed by #32586?

**Comment by steveklabnik** [comment]
It should have been, yes!
